pandas/core/case_when.py

```diff
@@ -0,0 +1,217 @@
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when((df.a.gt(0), df.a),   # condition, replacement
+    ...              (df.b.gt(0), df.b),   # condition, replacement
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+    from pandas._testing.asserters import assert_index_equal
+
+    # summary of logic -
+    # get booleans and replacements
+    # and iteratively get the final output via Series.mask
+    # iteration starts from the end,
+    # ensuring that if there are any booleans with the same replacement,
+    # the first one is taken
+    # as much as possible the computation is pushed to Series.mask
+    if not len(args):
+        raise ValueError(
+            "Kindly provide at least one boolean condition, "
+            "with a corresponding replacement."
+        )
+
+    for num, entry in enumerate(args):
+        if not isinstance(entry, tuple):
+            raise TypeError(
+                f"Argument{num} should be a tuple, instead, got {type(entry)}."
+            )
+        if len(entry) != 2:
+            raise ValueError(
+                f"Argument{num} should have a condition "
+                "and a replacement - it should have a length 2; "
+                f"however, it has a length {len(entry)}."
+            )
+    conditions, replacements = zip(*args)
+    common_dtype = [infer_dtype_from(replacement)[0] for replacement in replacements]
+
+    if default is not lib.no_default:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtype.append(arg_dtype)
+    else:
+        default = None
+    if len(set(common_dtype)) > 1:
+        common_dtype = find_common_type(common_dtype)
+        replacements = [
+            construct_1d_arraylike_from_scalar(
+                value=replacement, length=len(condition), dtype=common_dtype
+            )
+            if is_scalar(replacement)
+            else replacement.astype(common_dtype)
+            if isinstance(replacement, ABCNDFrame)
+            else pd_array(replacement, dtype=common_dtype)
+            for condition, replacement in zip(conditions, replacements)
+        ]
+        if (default is not None) and isinstance(default, ABCNDFrame):
+            default = default.astype(common_dtype)
+    else:
+        common_dtype = common_dtype[0]
+    if not isinstance(default, ABCNDFrame):
+        cond_indices = [cond for cond in conditions if isinstance(cond, ABCNDFrame)]
+        replacement_indices = [
+            replacement
+            for replacement in replacements
+            if isinstance(replacement, ABCNDFrame)
+        ]
+        cond_length = None
+        if replacement_indices:
+            for left, right in zip_longest(
+                replacement_indices, replacement_indices[1:]
+            ):
+                if right is None:
```

Comment:
Why not just use `zip` above? Then this condition is not needed.

Reply:
@rhshadrach which line are you referring?

Reply:
L150 is the condition I believe is not needed if you use `zip` on L147.

Reply:
sorry @rhshadrach I still do not your comment here. i'd appreciate a bit more guidance.

Reply:
You currently have:

```
            for left, right in zip_longest(
                replacement_indices, replacement_indices[1:]
            ):
                if right is None:
                    continue
                if assert_index_equal(left.index, right.index, check_order=False):
                    raise ValueError(
                        "All replacement objects must have the same index."
                    )
```

I think this is equivalent to

```
            for left, right in zip(
                replacement_indices, replacement_indices[1:]
            ):
                if assert_index_equal(left.index, right.index, check_order=False):
                    raise ValueError(
                        "All replacement objects must have the same index."
                    )
```